tensorboard/plugins/hparams/api.proto

```diff
@@ -0,0 +1,346 @@
+  string monitor_url = 9;
+}
+
+// Represents the status of a Session.
+enum Status {
+  STATUS_UNKNOWN = 0;
+  STATUS_SUCCESS = 1;
+  STATUS_FAILURE = 2;
+  STATUS_RUNNING = 3;
+}
+
+// Parameters for a ListSessionGroups API call.
+// See ListSessionGroups in http_api.md.
+// Since this proto is passed as a URL query parameter, care should be taken
+// when adding new fields to avoid exceeding the URL lenght limit.
+message ListSessionGroupsRequest {
+  // A list of ColParams messages--one for each "column" of a session group. A
+  // session group column contains either a metric evaluated at the current
+  // reported computation step or a hyperparameter value. The ColParams messages
+  // configure filtering and sorting of the resulting collection of
+  // session groups as follows:
+  //
+  // Filtering: each ColParams.filter field (see below) defines a subset of
+  // the domain of the associated column. Only session groups with each of
+  // their columns belonging to the corresponding subset will be returned.
+  // If no ColParams message is given for a column or if the filter field is
+  // unset, the column is regarded as unfiltered (i.e. the subset is taken as
+  // the entire column domain).
+  //
+  // Sorting: defined by the subset of ColParams messages whose 'order' field
+  // is not ORDER_UNSPECIFIED. In each of these the order field specifies the
+  // order (ascending or descending) of the values in the associated column
+  // of the session group. The order of the ColParams messages themselves
+  // determines the "significance" of the associated column as a sorting key:
+  // with the first being the primary sorting key, the second being
+  // the secondary sorting key, etc.
+  repeated ColParams col_params = 1;
+
+  // Fields controlling how to aggregate metrics across sessions within a
+  // session group.
+  // If metric_aggregation_type is AGGREGATION_AVG, each metric value of the
+  // session group is the average of the values of the metric across the
+  // sessions (at the same computation step).
+  // Otherwise, each metric value of the session group is the value of that
+  // metric for a "representative" session in the group. A representative
+  // session is a session having a "representative" value for the metric
+  // specified by the 'primary_metric' field below. The representative value
+  // for a metric is its minimum, maximum or median value across the sessions
+  // in the group, for metric_aggregation_type being AGGREGATION_MIN,
+  // AGGREGATION_MAX, or AGGREGATION_MEDIAN, respectively.
+  AggregationType aggregation_type = 2;
+  MetricName primary_metric = 3;
+
+  // Pagination support:
+  // The number of session groups in each page to return.
+  int32 page_size = 4;
+  // The token representing the page to return. Should be empty for the first
+  // page and copied from the 'next_page_token' field of the response for
+  // later pages.
+  string page_token = 5;
```

Comment:
I'm curious to know how you're thinking of implementing the pagination tokens, and I'm a little concerned that this won't quite be enough to make the pagination behave in ways that users expect.  A single token is good enough for an API case where the data is just one giant stream being consumed by an API client.  But in an actual user's pagination use cases, they will often want to navigate backwards as well as forwards, and potentially jump to arbitrary pages in the result set (i.e. jump to first page, last page, etc.).

Navigating backwards could be solved by keeping the old page data cached in the frontend (I think we have to cache it, since otherwise the query may have changed, resulting in pages that don't match what the user saw before, and potentially leading to missing results and/or incorrect page numbering.)  But jumping ahead in the result set is quite challenging I think, since if we don't cache all the intervening pages, navigating backwards might be broken (for the same reasons as just described), but if we do, this makes jumping to the last page equivalent to loading the entire result set which sort of defeats the point of paginating at the server side.

Maybe the right answer is just to not support any pagination options other than next page and previous page, and if the user wants to "jump to last" they have to manually reverse the sort order (it may be nice to provide a specific button for this that toggles the order on all active sorting filters).  Users may find this pretty annoying, but that might be acceptable if we enable it only for very high result set sizes (say 2000+ results).